core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -283,6 +283,44 @@ protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotI
     }
   }
 
+  /**
+   * Validates that no delete files matching a filter have been added to the table since a starting snapshot.
+   *
+   * @param base table metadata to validate
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param conflictDetectionFilter an expression used to find new conflicting delete files
+   * @param caseSensitive whether expression evaluation should be case sensitive
+   */
+  protected void validateAddedDeleteFiles(TableMetadata base, Long startingSnapshotId,
+                                          Expression conflictDetectionFilter, boolean caseSensitive) {
+    // if there is no current table state, no files have been added
+    if (base.currentSnapshot() == null) {
+      return;
+    }
+
+    Pair<List<ManifestFile>, Set<Long>> history =
+        validationHistory(base, startingSnapshotId, VALIDATE_ADDED_DELETE_FILES_OPERATIONS, ManifestContent.DELETES);
+    List<ManifestFile> deleteManifests = history.first();
+
+    DeleteFileIndex.Builder deleteIndexBuilder = DeleteFileIndex.builderFor(ops.io(), deleteManifests)
+        .filterData(conflictDetectionFilter)
+        .caseSensitive(caseSensitive)
+        .specsById(ops.current().specsById());
+
+    // the starting snapshot could have been expired concurrently
+    if (startingSnapshotId != null && base.snapshot(startingSnapshotId) != null) {
+      Snapshot startingSnapshot = base.snapshot(startingSnapshotId);
+      long startingSequenceNumber = startingSnapshot.sequenceNumber();
+      deleteIndexBuilder.afterSequenceNumber(startingSequenceNumber);
+    }
+
+    DeleteFileIndex deletes = deleteIndexBuilder.build();
+
+    ValidationException.check(deletes.isEmpty(),
+        "Found new conflicting delete files that can apply to records matching %s",
```

Comment:
For debugging some complex concurrency use case, we find the list of data files in exception message very helpful to identify where it's coming from.  Is a similar thing possible with delete files (or at least print a sample of them)?

Reply:
It is not something possible right away but I can try to expose a list of delete files from the delete file index.
I agree that will be useful. I'll update.

Reply:
Yeah, I agree here. It would be great to get the list.

Reply:
Added a list. Could you check, @szehon-ho?

Reply:
Resolving as this has been addressed.